pkg/controller/utils/elasticsearch.go

```diff
@@ -62,3 +104,265 @@ func GetElasticsearchClusterConfig(ctx context.Context, cli client.Client) (*ren
 
 	return render.NewElasticsearchClusterConfigFromConfigMap(configMap)
 }
+
+type ElasticClient interface {
+	SetILMPolicies(client.Client, context.Context, *operatorv1.LogStorage) error
+}
+
+type esClient struct {
+	client *elastic.Client
+	lock   sync.Mutex
+}
+
+func NewElasticClient() ElasticClient {
+	return &esClient{}
+}
+
+// SetILMPolicies creates ILM policies for each timeseries based index using the retention period and storage size in LogStorage
+func (es *esClient) SetILMPolicies(client client.Client, ctx context.Context, ls *operatorv1.LogStorage) error {
+	if err := es.createElasticClient(client, ctx); err != nil {
+		return err
+	}
+	policyList := es.listILMPolicies(ls)
+	return es.createOrUpdatePolicies(ctx, policyList)
+}
+
+func (es *esClient) createElasticClient(client client.Client, ctx context.Context) error {
+	if es.client == nil {
+		es.lock.Lock()
```

Comment:
This lock should be before accessing the `es.client`, as it is it would be possible that two calls could happen to createElasticClient and if they both do the if check before the lock then they would both create the client.